README.md
# node-red-contrib-iiot-rpi-ds18b20

[![platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
[![platform](https://img.shields.io/badge/platform-Raspberry--Pi-ff69b4)](https://www.raspberrypi.com/)
![NPM version](https://badge.fury.io/js/node-red-contrib-iiot-ds18b20.svg)
![NPM](https://img.shields.io/npm/l/node-red-contrib-iiot-ds18b20)
[![Donate](https://img.shields.io/badge/Donate-PayPal-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZDRCZBQFWV3A6)

A Node-RED node for multiple ds18b20 digital temperature sensor<br>

## Installation
Install with Node-Red Palette Manager or npm command:
```
cd ~/.node-red
npm install node-red-contrib-iiot-ds18b20
```

## Usage
- Multiple sensors are read on cycle.
- To prevents locks only one node is allowed.
- The output values are °C or °F and send as number array.
- This node works on Raspberry Pi with 32bit or 64bit OS.
- Enable 1-Wire with raspi-config.

[CHANGELOG](CHANGELOG.md)

## Set sensor resolution
You can change the sensor resolution with tool in subfolder **setSensor**.<br>
File **readme.txt** describes how the tool is used.<br>

## Example
This example reads temperatures from 3 sensors.<br>
Values are in °C.<br>
Each sensor value is send to outputs 0..2 in function node.<br>
In Node-Red Dashboard, these are displayed in gauges.<br>
Each sensor value is compared with fix value and send to outputs 3..5 in function node.<br>
In Node-Red Dashboard, these are displayed in sliders as alarm.<br>

### JavaScript Code in Function Node:
```javascript
return [
    { payload: msg.payload[0] }, // output 1
    { payload: msg.payload[1] }, // output 2
    { payload: msg.payload[2] }, // output 3  
    { payload: msg.payload[0] >= 30.5 }, // output 4
    { payload: msg.payload[1] >= 20.5 }, // output 5
    { payload: msg.payload[2] >= 19.3 } // output 6  
];
```

![image info](images/temp_alarm.png)



CHANGELOG.md
2.0.1
- Initial Release

2.0.2
- bug fix, wrong github path

2.0.3
- bug fix, change update time from ms to s
- add sensor list to dialog

2.0.4
- bug fix, wrong path in README.md
- Init array with -55°C or -67°F
- Change topic array string to topic string

2.0.5
- Add as present for Node-Red user a tool for set sensor resolution

2.0.6
- bug fix, setsensor don't release resource (memory leak)
- Add payload index number to sensor list in dialog
- Change max. sensor count from 32 to 64 (has anyone connected so many sensors?)

setSensor/syslib.js
/**
 * Copyright 2023 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use node file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

const fs = require('fs');

/**
* Check for Raspberry Pi and OS.
* @returns module extension depends on OS
*/
function getRpiTarget()
{
    if (process.platform !== 'linux')
        return undefined;
    
    var cpuinfo = fs.readFileSync("/proc/cpuinfo").toString();

    if (cpuinfo.indexOf(": BCM") === -1)
        return undefined;
    
    if (process.arch === "arm64")
        return "_s64.node";
    else if (process.arch === "arm")
        return "_s32.node";
    
    return undefined;
}

/**
* Check for Raspberry Pi and loads module.
* @param module - the name name
* @returns module object
*/
module.exports.LoadModule = function(module) {
    try {
        var modName = getRpiTarget();
        if (modName !== undefined)
            return require("./" + module + modName);
    }
    catch (e) { console.log(e); }
    return undefined;
}


setSensor/setsensor.js
/**
 * Copyright 2023 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

const syslib = require("./syslib.js");
const sysmodule = syslib.LoadModule("rpi_ds18b20_set");

if (sysmodule === undefined) {
    console.log("driver not load, wrong os or not Raspi");
    process.exit(1);
}

// default values if parameter not given
var pin = 4;
var resolution = 12;

// call:
// node setsensor [resolution] [pin]

switch(process.argv.length) {
    case 4: // resolution and pin given
        pin = Number(process.argv[3]);
        if (isNaN(pin) || (pin < 2) || (pin > 27)) {
            console.log("pin must 2..27");
            process.exit(1);
        }
    // no break
    case 3: // resolution given
        resolution = Number(process.argv[2]);
        if (isNaN(resolution) || (resolution < 9) || (resolution > 12)) {
            console.log("resolution must 9..12");
            process.exit(1);
        }
}

// call magic function
if (!sysmodule.changeSensorsResolution(pin, resolution)) {
    console.log("error on change resolution");
    process.exit(1);
}

console.log("resolution changed to ", resolution, " on pin ", pin);


iiot-ds18b20-fs.js
/**
 * Copyright 2023 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
    const syslib = require("./lib/syslib.js");
    const sysmodule = syslib.LoadModule("rpi_ds18b20");

    RED.nodes.registerType("iiot-ds18b20-fs", function(n) {
        var node = this;
        RED.nodes.createNode(node, n);

        node.tupdate = n.tupdate;
        node.fahrenheit = n.fahrenheit;
        node.tofix = n.tofix;

        node.name = "iiot-ds18b20-fs";
        node.onwork = false;
        node.iserror = false;
        node.ndevice = 0;
        node.ch_cnt = 0;
        node.status_txt = "↻ " + node.tupdate + "s ";

        if (sysmodule === undefined)
            node.iserror = syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
        else if (sysmodule.isused())
            node.iserror = syslib.outError(node, "already used", "node already used");
        else if (!sysmodule.init_fs())
            node.iserror = syslib.outError(node, "init error", "init error, no 1-wire setup");
        else {
            node.ndevice = sysmodule.get_ndevice_fs();

            if (node.ndevice == 0)
                node.iserror = syslib.setStatus(node, node.status_txt + " nodevice", "grey");
            else {
                node.ctxvar = new Array(node.ndevice).fill(node.fahrenheit ? -67 : -55);
                syslib.setStatus(node, node.status_txt, "grey");
            }
        }

        function update(index) {
            sysmodule.get_temperatur_fs(index, function(readval, index, device_id) {
                if (readval === null)
                    node.read_error = syslib.outError(node, "read error: " + device_id);
                else {
                    var val_read = node.fahrenheit ? (readval * (9/5)) + 32 : readval;
                    val_read = syslib.toFixed(val_read, node.tofix);

                    if (node.ctxvar[index] !== val_read) {
                        node.ctxvar[index] = val_read;
                        node.send({ payload: node.ctxvar, topic: device_id });
                    }
                }

                node.ch_cnt++;

                if (node.ch_cnt >= node.ndevice)
                {
                    if (!node.read_error)
                        syslib.setStatus(node, node.status_txt);

                    node.onwork = false;
                }
            });
        }

        if (!node.iserror)
            node.id_interval = setInterval(function() {
                if (node.onwork)
                    return;

                node.onwork = true;
                node.ch_cnt = 0;
                node.read_error = false;

                for ( var index = 0; index < node.ndevice; index++)
                    update(index);
            }, node.tupdate * 1000);

        node.on('close', function () {
            clearInterval(node.id_interval);
            sysmodule.deinit();
        });
    });

    RED.httpAdmin.get('/sensorlist', function(req, res) {
        res.send(sysmodule.get_list_fs());
    });
}
